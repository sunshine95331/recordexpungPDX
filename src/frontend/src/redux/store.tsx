// See the following guides for an explanation:
// https://redux-starter-kit.js.org/usage/usage-guide
// https://redux.js.org/recipes/usage-with-typescript
import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import { combineReducers } from 'redux';

// Reducers:
import { searchReducer } from './search/reducer';

const rootReducer = combineReducers({
  search: searchReducer,
});

const store = configureStore({
  reducer: rootReducer,
  middleware: [...getDefaultMiddleware()],
});

export type AppState = ReturnType<typeof rootReducer>;

export default store;
