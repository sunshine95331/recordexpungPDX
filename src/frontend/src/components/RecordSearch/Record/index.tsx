import React, { useState } from "react";
import { useAppSelector, useAppDispatch } from "../../../redux/hooks";
import {
  startEditing,
  doneEditing,
  selectIsEditing,
} from "../../../redux/editingSlice";
import Cases from "./Cases";
import Case from "./Case";
import AddButton from "./AddButton";
import Link from "../../common/Link";

export default function Record() {
  const dispatch = useAppDispatch();
  const record = useAppSelector((state) => state.search.record);
  const isEditing = useAppSelector(selectIsEditing);
  const [enableEditing, setEnableEditing] = useState(false);
  const [addingNewCase, setAddingNewCase] = useState(false);
  const [nextNewCaseNum, setNextNewCaseNum] = useState(1);

  const createNextBlankCase = () => {
    return {
      balance_due: 0,
      birth_year: 0,
      case_detail_link: "",
      case_number: "CASE-" + ("000" + nextNewCaseNum).slice(-4),
      charges: [],
      citation_number: "",
      current_status: "",
      date: "",
      location: "",
      name: "",
      violation_type: "",
      district_attorney_number: "",
      edit_status: "ADD",
    };
  };

  const handleAddCaseClick = () => {
    dispatch(startEditing());
    setAddingNewCase(true);
  };

  return (
    <>
      {addingNewCase && (
        <div className="bg-gray-blue-2 shadow br3 overflow-auto mb3">
          <Case
            whenEditing={() => {
              dispatch(startEditing());
            }}
            whenDoneEditing={() => {
              dispatch(doneEditing());
              setAddingNewCase(false);
              setNextNewCaseNum(nextNewCaseNum + 1);
            }}
            case={createNextBlankCase()}
            editing={true}
            isNewCase={true}
            showEditButtons={!isEditing && enableEditing}
            customElementId="new-case"
          />
        </div>
      )}
      <div className="tr">
        {enableEditing ? (
          !isEditing && (
            <>
              <AddButton
                onClick={handleAddCaseClick}
                actionName={"Add"}
                text={"Case"}
              />
              <div className="pb3" />
            </>
          )
        ) : (
          <>
            <button
              className="inline-flex bg-white f6 fw5 br2 ba b--black-10 mid-gray link hover-blue pv1 ph2 mb3"
              onClick={() => {
                setEnableEditing(true);
              }}
            >
              Enable Editing
            </button>{" "}
            <Link
              to="/manual#editing"
              className=" gray link hover-blue underline"
              iconClassName="fas fa-question-circle link hover-dark-blue gray"
              hiddenText="Editing Help"
            />
          </>
        )}
      </div>

      {record?.cases && (
        <Cases
          cases={record.cases}
          showEditButtons={!isEditing && enableEditing}
          whenEditing={() => {
            dispatch(startEditing());
          }}
          whenDoneEditing={() => {
            dispatch(doneEditing());
          }}
        />
      )}
    </>
  );
}
