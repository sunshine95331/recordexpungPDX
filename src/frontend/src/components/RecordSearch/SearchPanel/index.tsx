import React, { useEffect, useState } from "react";
import moment from "moment";
import { AliasData, AliasFieldNames } from "./types";
import { useAppSelector, useAppDispatch } from "../../../redux/hooks";
import {
  selectSearchFormAliases,
  selectSearchFormDate,
  setSearchFormDate,
  updateSearchFormAlias,
  addSearchFormAlias,
  removeSearchFormAlias,
} from "../../../redux/searchFormSlice";
import {
  isValidOptionalWildcard,
  areAnyBlank,
  isValidOptionalDate,
} from "./validators";
import { searchRecord } from "../../../redux/search/actions";
import Alias from "./Alias";
import Field from "./Field";
import IconButton from "../../common/IconButton";
import InvalidInputs from "../../InvalidInputs";

export default function SearchPanel() {
  const dispatch = useAppDispatch();
  const aliases = useAppSelector(selectSearchFormAliases);
  const date = useAppSelector(selectSearchFormDate);
  const [errorMessages, setErorrMessages] = useState<string[]>([]);

  useEffect(() => {
    if (date === "") dispatch(setSearchFormDate(moment().format("M/D/YYYY")));
  }, [date, dispatch]);

  // A map of error messages and validators that return true if messages should be shown
  const validators = {
    "First and last name are required.": (alias: AliasData) =>
      areAnyBlank(alias.first_name, alias.last_name),

    "The date format must be MM/DD/YYYY.": (alias: AliasData) =>
      !isValidOptionalDate(alias.birth_date),

    "A wildcard in First Name field must be at the end and follow at least one letter.":
      (alias: AliasData) => !isValidOptionalWildcard(alias.first_name, 2),

    "A wildcard in the Last Name field must be at the end and follow at least two letters.":
      (alias: AliasData) => !isValidOptionalWildcard(alias.last_name, 3),
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    let messages = Object.entries(validators).reduce(
      (errors, [message, isInvalid]) => {
        if (aliases.some(isInvalid)) errors.push(message);
        return errors;
      },
      [] as string[]
    );

    setErorrMessages(messages);

    if (messages.length === 0) dispatch(searchRecord());
  };

  return (
    <form className="mw7 center" onSubmit={handleSubmit} noValidate>
      <div className="flex">
        <Field
          coda="mm/dd/yyyy"
          name="today"
          label="Expunge Date"
          content={date}
          divMarkup="pl2-r"
          onChange={(newDate: string) => {
            dispatch(setSearchFormDate(newDate));
          }}
          required={true}
          errorMessage="today_msg"
        />
      </div>

      {aliases.map((alias: AliasData, index: number) => {
        return (
          <React.Fragment key={index}>
            {index > 0 && (
              <hr className="bb b--black-05 mt2 mt3-ns mb3 mb4-ns" />
            )}
            <Alias
              ind={index}
              aliasData={alias}
              onChange={(attribute: AliasFieldNames, value: string) => {
                dispatch(updateSearchFormAlias({ index, attribute, value }));
              }}
              onRemoveClick={() => {
                dispatch(removeSearchFormAlias(index));
              }}
              hideRemoveButton={aliases.length === 1}
            />
          </React.Fragment>
        );
      })}

      <div className="flex">
        <IconButton
          styling="blank"
          displayText="Alias"
          buttonClassName="w4 tc br2 bg-gray-blue-2 link hover-dark-blue mid-gray fw5 pv3 ph3 mr2"
          iconClassName="fa-plus-circle pr1"
          onClick={() => dispatch(addSearchFormAlias())}
        />

        <IconButton
          styling="blank"
          type="submit"
          displayText="Search"
          buttonClassName="fw7 br2 bg-blue white bg-animate hover-bg-dark-blue db w-100 tc pv3 btn--search"
          iconClassName="fa-search pr2"
        />
      </div>

      <InvalidInputs contents={errorMessages} />
    </form>
  );
}
