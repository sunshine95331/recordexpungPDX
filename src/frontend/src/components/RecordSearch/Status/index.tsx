import React from "react";
import { useAppSelector } from "../../../redux/hooks";
import LoadingSpinner from "../../LoadingSpinner";

export default function Status() {
  const record = useAppSelector((state) => state.search.record);
  const loading = useAppSelector((state) => state.search.loading);

  return (
    <section>
      {loading === "loading" ? (
        <LoadingSpinner inputString="your search results" />
      ) : record?.cases?.length === 0 ? (
        <p className="bg-light-gray mv4 pa4 br3 fw6">
          No search results found.
        </p>
      ) : null}
    </section>
  );
}
