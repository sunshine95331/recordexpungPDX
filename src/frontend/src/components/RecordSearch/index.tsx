import React, { useState, useEffect } from "react";
import { useAppDispatch, useAppSelector } from "../../redux/hooks";
import history from "../../service/history";
import { stopDemo } from "../../redux/search/actions";
import { hasOeciToken } from "../../service/cookie-service";
import useSetupPage from "../../hooks/useSetupPage";
import SearchPanel from "./SearchPanel";
import Record from "./Record";
import Status from "./Status";
import Assumptions from "./Assumptions";

export default function RecordSearch() {
  const [shouldDisplay, setShouldDisplay] = useState(false);
  const record = useAppSelector((state) => state.search.record);
  const dispatch = useAppDispatch();

  useSetupPage("Search Records");

  useEffect(() => {
    dispatch(stopDemo());

    if (!hasOeciToken()) {
      history.replace("/oeci");
    } else {
      setShouldDisplay(true);
    }
  }, [shouldDisplay, dispatch]);

  return shouldDisplay ? (
    <main className="mw8 center f6 f5-l ph2">
      <SearchPanel />
      <Status record={record} />
      <Record record={record} />
      <Assumptions />
    </main>
  ) : (
    <></>
  );
}
