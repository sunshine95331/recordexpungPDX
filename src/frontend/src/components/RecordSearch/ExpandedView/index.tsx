import { useAppDispatch } from "../../../redux/hooks";
import { startLoadingSummary } from "../../../redux/summarySlice";
import { buildAndSendDownloadPdfRequest } from "../../../redux/search/actions";
import { clearAllData } from "../../../redux/store";
import CasesSummary from "../ExpandedView/CasesSummary";
import Stats from "../ExpandedView/SummaryStats";
import Record from "../Record";
import CasesNavList from "../ExpandedView/CasesNav";
import AssumptionsNew from "../ExpandedView/AssumptionsNew";
import { IconButton } from "../../common/IconButton";
import ExpungementFormsInfo from "./ExpungementFormsInfo";
import UserDataForm from "../../FillForms/UserDataForm";
import { panelClass, mainWrapper, singlePanelClass } from "../Layout";

export const headingLargeClass = "f3 fw8 tc pt2 mb4 ph4";
const wrapperClass = "flex-l mh2 mt3";
const panelHeadingClass = "f5 fw7 tc pv3 ";
const leftPanelClass = "w-70-l f6 ph4 mr3";
const rightPanelClass = "w-30-l pl3 pb2 ";

export default function ExpandedView({ showColor }: { showColor: boolean }) {
  const dispatch = useAppDispatch();

  function handleDownloadSummaryClick() {
    dispatch(startLoadingSummary());
    dispatch(buildAndSendDownloadPdfRequest);
  }

  function handleStartOverClick() {
    dispatch(clearAllData());
    window.scroll(0, 0);
  }

  return (
    <>
      <section className={wrapperClass}>
        <div className={panelClass + leftPanelClass}>
          <h2 className={panelHeadingClass}>Review Summary</h2>
          <CasesSummary showColor={showColor} />
        </div>

        <div className={panelClass + rightPanelClass}>
          <h3 className={panelHeadingClass}>Counts</h3>
          <Stats showColor={showColor} />
        </div>
      </section>

      <section className={wrapperClass}>
        <div className={panelClass + leftPanelClass}>
          <h2 className={panelHeadingClass}>Analyze Cases</h2>
          <Record />
        </div>

        <div className={panelClass + rightPanelClass + "self-start sticky-l"}>
          <h3 className={panelHeadingClass}>Quick Links</h3>
          <CasesNavList className="f6 overflow-y-auto vh-75" />
        </div>
      </section>

      <section className={mainWrapper + panelClass + singlePanelClass + "tc"}>
        <h2 className={headingLargeClass}>
          Get The Expungement Analysis Report
        </h2>
        <IconButton
          type="button"
          buttonClassName="bg-blue white hover-bg-dark-blue"
          iconClassName="fa-download pr2"
          displayText="Download Now"
          onClick={handleDownloadSummaryClick}
        />
      </section>

      <section className={mainWrapper + panelClass + singlePanelClass}>
        <h2 className={headingLargeClass}>Verify Assumptions</h2>
        <AssumptionsNew className="f5 lh-copy mw6-ns mw5 pl4-ns center list" />
      </section>

      <section
        className={mainWrapper + panelClass + singlePanelClass + "flex-l"}
      >
        <div className="w-60-l">
          <h2 className={headingLargeClass}>Expungement Forms</h2>
          <ExpungementFormsInfo className="f5 lh-copy mw6-ns mw5 pl4-ns center list" />
        </div>

        <div className="w-40-l">
          <UserDataForm />
        </div>
      </section>

      <section
        className={mainWrapper + panelClass + singlePanelClass + "tc pb5 mb3"}
      >
        <h2 className={headingLargeClass}>Finish Up</h2>
        <IconButton
          type="button"
          buttonClassName="bg-red white hover-bg-dark-red "
          iconClassName="fa-arrow-up pr2"
          displayText="Clear Data And Start Over"
          onClick={handleStartOverClick}
        />
      </section>
    </>
  );
}
