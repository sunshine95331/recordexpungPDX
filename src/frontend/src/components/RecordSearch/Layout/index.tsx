import React, { useState, useEffect } from "react";
import { useAppSelector } from "../../../redux/hooks";
import { selectStats } from "../../../redux/statsSlice";
import useRadioGroup from "../../../hooks/useRadioGroup";
import setUpScrollSpy from "../ExpandedView/setupScrollSpy";
import SearchPanel from "../SearchPanel";
import Status from "../Status";
import RecordSummary from "../Record/RecordSummary";
import Record from "../Record";
import Assumptions from "../Assumptions";
import { convertCaseNumberIntoLinks } from "../Record/util";
import ViewOptions from "../ExpandedView/ViewOptions";
import ExpandedView from "../ExpandedView";

function ErrorMessage({ message, idx }: { message: string; idx: number }) {
  const id = "record_error_" + idx;
  return (
    <p role="status" id={id} key={id} className="bg-washed-red mv3 pa3 br3 fw6">
      {convertCaseNumberIntoLinks(message)}
    </p>
  );
}

export const mainWrapper = "mw8 center f6 f5-l ";
export const panelClass = "bg-white shadow br3 ";
export const singlePanelClass = "pa4 mt3 ";

export default function Layout({
  topSection,
}: {
  topSection?: React.ReactElement;
}) {
  const record = useAppSelector((state) => state.search.record);
  const stats = useAppSelector(selectStats);
  const [showColor, setShowColor] = useState(true);
  const { selectedRadioValue, ...radioGroupProps } = useRadioGroup({
    label: "Summary overview sort options",
    initialValue: "Default",
  });
  const radioGroupOptions = ["Default", "Expanded"];
  const isExpandedView = selectedRadioValue === "Expanded";
  useEffect(setUpScrollSpy, [isExpandedView, record]);

  useEffect(() => {
    document.querySelector("html")?.classList.toggle("smooth-scroll");
    return () => {
      document.querySelector("html")?.classList.toggle("smooth-scroll");
    };
  });

  return (
    <main>
      <div className={mainWrapper + "ph2"}>
        {topSection}
        {!topSection && <h1 className="visually-hidden">Record Search</h1>}

        <section className={panelClass + singlePanelClass}>
          <SearchPanel />
        </section>

        <section className={panelClass + "mv3 ph3 pv1"}>
          <div className="flex flex-wrap items-center">
            <div className="f5 fw7 mr4">View Options</div>

            <ViewOptions
              optionLabels={radioGroupOptions}
              radioGroupProps={radioGroupProps}
              showColor={showColor}
              isExpandedView={isExpandedView}
              handleShowColorChange={() => setShowColor(!showColor)}
            />
          </div>
        </section>

        <Status />

        {record?.errors?.map((message: string, idx: number) => (
          <ErrorMessage key={idx} message={message} idx={idx} />
        ))}

        {!isExpandedView && (
          <>
            <section>
              <RecordSummary />
              <Record />
            </section>

            <Assumptions />
          </>
        )}
      </div>

      {isExpandedView && stats.totalCases > 0 && (
        <ExpandedView showColor={showColor} />
      )}
    </main>
  );
}
