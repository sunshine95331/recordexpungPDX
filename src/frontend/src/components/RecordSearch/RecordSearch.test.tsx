import React from "react";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import store from "../../redux/store";
import history from "../../service/history";
import RecordSearch from ".";

const mockHasOeciToeken = jest.fn();

jest.mock("../../redux/search/initialState", () => {
  let state = jest.requireActual("../../redux/search/initialState").default;
  state.record = jest.requireActual("../../data/demo/johnCommon").default;
  return state;
});

jest.mock("../../service/cookie-service", () => ({
  ...jest.requireActual("../../service/cookie-service"),
  hasOeciToken: () => mockHasOeciToeken(),
}));

function doRender() {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <RecordSearch />
      </MemoryRouter>
    </Provider>
  );
}

describe("when logged in", () => {
  beforeEach(() => {
    mockHasOeciToeken.mockReturnValue(true);
  });

  it("renders correctly with John Common demo record", () => {
    const RealDate = global.Date;
    const mockDate = new Date(2023, 0, 11);
    // @ts-ignore
    const spy = jest.spyOn(global, "Date").mockImplementation(() => mockDate);
    // @ts-ignore
    spy.now = RealDate.now;

    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <RecordSearch />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();

    global.Date = RealDate;
  });

  it("dispatches the stop demo action", () => {
    const dispatchSpy = jest.spyOn(store, "dispatch");

    doRender();
    expect(dispatchSpy).toHaveBeenCalledWith({
      type: "STOP_DEMO",
    });
  });

  it("displays the correct document title", () => {
    doRender();
    expect(global.window.document.title).toBe("Search Records - RecordSponge");
  });

  it("does not redirect", () => {
    const historyReplaceSpy = jest.spyOn(history, "replace");

    doRender();
    expect(historyReplaceSpy).not.toHaveBeenCalled();
  });
});

describe("when not logged in", () => {
  beforeEach(() => {
    mockHasOeciToeken.mockReturnValue(false);
  });

  it("dispatches the stop demo action", () => {
    const dispatchSpy = jest.spyOn(store, "dispatch");

    doRender();
    expect(dispatchSpy).toHaveBeenCalledWith({
      type: "STOP_DEMO",
    });
  });

  it("redirects", () => {
    const historyReplaceSpy = jest.spyOn(history, "replace");

    doRender();
    expect(historyReplaceSpy).toHaveBeenCalledWith("/oeci");
  });

  it("does not display content", () => {
    doRender();
    expect(screen.queryByText(/date/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/search/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/case/i)).not.toBeInTheDocument();
  });
});
