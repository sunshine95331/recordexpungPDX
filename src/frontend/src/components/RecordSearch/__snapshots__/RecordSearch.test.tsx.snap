// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when logged in renders correctly with John Common demo record 1`] = `null`;
